amp/system-oracle/workaround-non-cdb-2.4-upgrade/database.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+require 'active_support/string_inquirer'
+require 'system/database/triggers'
+require 'system/database/procedures'
+module System
+  module Database
+    class ConnectionError < ActiveRecord::NoDatabaseError; end
+    module_function
+
+    def configuration_specification
+      configurations = Rails.application.config.database_configuration
+      resolver = ActiveRecord::ConnectionAdapters::ConnectionSpecification::Resolver.new(configurations)
+      spec = ActiveRecord::ConnectionHandling::DEFAULT_ENV.call.to_sym
+
+      resolver.spec(spec)
+    end
+
+    def adapter_method
+      ActiveSupport::StringInquirer.new(configuration_specification.adapter_method)
+    end
+
+    def oracle?
+      adapter_method.oracle_enhanced_connection?
+    end
+
+    def mysql?
+      adapter_method.mysql2_connection?
+    end
+
+    # Just adding another connection to the pool so we do not mess up with the primary connection
+    # And just forget about it after
+    class ConnectionProbe < ActiveRecord::Base
+      extend System::Database
+      self.table_name = 'accounts'
+
+      class << self
+        def connection_config
+          spec = configuration_specification.config.dup
+          if oracle?
+            spec[:password] = ENV.fetch('ORACLE_SYSTEM_PASSWORD') {|key| raise KeyError, "Environment #{key} is mandatory"}
+            spec[:username] = 'SYSTEM'
+          end
+          spec
+        end
+
+        def ready?
+          pool = establish_connection connection_config
+          result = nil
+          pool.with_connection do |connection|
+            result = connection.select_value(sql_for_readiness).tap { pool.disconnect! }
+          end
+          result.to_s == '1'
+        end
+
+        def sql_for_readiness
+          if oracle?
+            "SELECT 1 FROM DUAL"
```

Comment:
Probably better to modify a bit https://github.com/3scale/porta/blob/2.4.0/lib/system/database.rb#L58

by

```ruby
 "SELECT 1 FROM v$pdbs WHERE name COLLATE BINARY_CI = '#{connection_config[:database]}' AND open_mode = 'READ WRITE'"
```

select open_mode from v$database

Reply:
Not tested and not enough time for that. :/

Reply:
https://github.com/3scale/porta/pull/766